srml/generic-asset/src/lib.rs

```diff
@@ -0,0 +1,1241 @@
+//! 	// ...
+//! 	T::Currency::deposit_into_existing(transactor, amount)?;
+//! 	// ...
+//! 	Ok(())
+//! }
+//!
+//! # fn main() {}
+//! ```
+//!
+//! ## Genesis config
+//!
+//! The Generic Asset module depends on the [`GenesisConfig`](./struct.GenesisConfig.html).
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use parity_codec::{Decode, Encode, HasCompact};
+
+use primitives::traits::{
+	CheckedAdd, CheckedSub, MaybeSerializeDebug, Member, One, Saturating, SimpleArithmetic, Zero,
+};
+
+use rstd::prelude::*;
+use rstd::{cmp, result};
+use support::dispatch::Result;
+use support::{
+	decl_event, decl_module, decl_storage, ensure,
+	traits::{
+		Currency, ExistenceRequirement, Imbalance, LockIdentifier, LockableCurrency, ReservableCurrency,
+		SignedImbalance, UpdateBalanceOutcome, WithdrawReason, WithdrawReasons,
+	},
+	Parameter, StorageDoubleMap, StorageMap, StorageValue,
+};
+use system::ensure_signed;
+
+mod mock;
+mod tests;
+
+pub use self::imbalances::{NegativeImbalance, PositiveImbalance};
+
+pub trait Trait: system::Trait {
+	type Balance: Parameter
+		+ Member
+		+ SimpleArithmetic
+		+ Default
+		+ Copy
+		+ MaybeSerializeDebug;
+	type AssetId: Parameter + Member + SimpleArithmetic + Default + Copy;
+	type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+}
+
+pub trait Subtrait: system::Trait {
+	type Balance: Parameter
+		+ Member
+		+ SimpleArithmetic
+		+ Default
+		+ Copy
+		+ MaybeSerializeDebug;
+	type AssetId: Parameter + Member + SimpleArithmetic + Default + Copy;
+}
+
+impl<T: Trait> Subtrait for T {
+	type Balance = T::Balance;
+	type AssetId = T::AssetId;
+}
+
+/// Asset creation options.
+#[cfg_attr(feature = "std", derive(Debug))]
+#[derive(Clone, Encode, Decode, PartialEq, Eq)]
+pub struct AssetOptions<Balance: HasCompact, AccountId> {
+	/// Initial issuance of this asset. All deposit to the creater of the asset.
+	#[codec(compact)]
+	pub initial_issuance: Balance,
+	/// Which accounts are allowed to possess this asset.
+	pub permissions: PermissionLatest<AccountId>,
+}
+
+/// Owner of an asset.
+#[cfg_attr(feature = "std", derive(Debug))]
+#[derive(Clone, Encode, Decode, PartialEq, Eq)]
+pub enum Owner<AccountId> {
+	/// No owner.
+	None,
+	/// Owned by an AccountId
+	Address(AccountId),
+}
+
+impl<AccountId> Default for Owner<AccountId> {
+	fn default() -> Self {
+		Owner::None
+	}
+}
+
+/// Asset permissions
+#[cfg_attr(feature = "std", derive(Debug))]
+#[derive(Clone, Encode, Decode, PartialEq, Eq)]
+pub struct PermissionsV1<AccountId> {
+	/// Who have permission to update asset permission
+	pub update: Owner<AccountId>,
+	/// Who have permission to mint new asset
+	pub mint: Owner<AccountId>,
+	/// Who have permission to burn asset
+	pub burn: Owner<AccountId>,
+}
+
+/// Versioned asset permission
+#[cfg_attr(feature = "std", derive(Debug))]
+#[derive(Clone, Encode, Decode, PartialEq, Eq)]
+pub enum PermissionVersions<AccountId> {
```

Comment:
seeing how this is going to be used, it's safer to make it `#[repr(u8)]` and explicitly set the `V1` item to a value (e.g. `0`).

Reply:
You may have misunderstand the intent usage of this struct. Let's say V2 add a `selfdestruct` permission, it will be like:

```
pub struct PermissionsV1<AccountId> {
	pub update: Owner<AccountId>,
	pub mint: Owner<AccountId>,
	pub burn: Owner<AccountId>,
}

pub struct PermissionsV2<AccountId> {
	pub update: Owner<AccountId>,
	pub mint: Owner<AccountId>,
	pub burn: Owner<AccountId>,
	pub selfdestruct: Owner<AccountId>,
}

/// Versioned asset permission
#[cfg_attr(feature = "std", derive(Debug))]
#[derive(Clone, Encode, Decode, PartialEq, Eq)]
pub enum PermissionVersions<AccountId> {
	V1(PermissionsV1<AccountId>),
	V2(PermissionsV2<AccountId>),
}

/// Alias to latest asset permissions
pub type PermissionLatest<AccountId> = PermissionsV2<AccountId>;

```

So that `PermissionVersions` is not a version number, but an enum can holds all the possible versions.
With storage
`pub Permissions get(get_permission): map T::AssetId => PermissionVersions<T::AccountId>;`

 different permission versions can be stored in a backward compatible way.

With `impl<AccountId> Into<PermissionLatest<AccountId>> for PermissionVersions<AccountId>`
the code only need to deal with latest version and the `Into` implementation should take care of the migration

Reply:
yeah, but afaik there's no guarantee in Rust or `parity-codec` that `FooV1::Bar(42)` and `FooV2::Bar(42)` will `Encode` to the same value:

```
enum FooV1 { ..., Bar(u8) }
enum FooV2 { ..., Bar(u8), Baz(u32) }
```

Reply:
SCALE encoding format should guarantee that? no? 
the encoding format for `parity-codec` should be well defined (per SCALE format spec) so it should be safe to make this assumption.

Reply:
It will probably work as it is anyway for now, but there is no guarantee unless you explicitly tell the codec that these of these items should encode to a specific integer value.

Reply:
I have manually implemented Encode / Decode to make it more explicit